fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/query/plan/synthetic/SyntheticRecordPlannerTest.java

```diff
@@ -825,6 +831,363 @@ public void aggregateJoinIndex() throws Exception {
+        );
+
+        // Add an index on the cc field so that the join planner can use the index to resolve join pairs
+        metaDataBuilder.addIndex(metaDataBuilder.getRecordType("OrderWithHeader"),
+                new Index("OrderWithHeader$cc", concat(field("___header").nest("z_key"), field("cc", KeyExpression.FanType.FanOut).nest("string_value"))));
+        // Add a join index listing all of the CC'd customer names for a given order number
+        metaDataBuilder.addIndex(joined, new Index("OrderCCNames",
+                concat(field("order").nest(field("___header").nest("z_key")), field("order").nest("order_no"), field("cust").nest("name"))));
+
+        List<TestRecordsJoinIndexProto.CustomerWithHeader> customers = IntStream.range(0, 10)
+                .mapToObj(i -> TestRecordsJoinIndexProto.CustomerWithHeader.newBuilder()
+                        .setHeader(TestRecordsJoinIndexProto.Header.newBuilder().setZKey(1L).setIntRecId(i).build())
+                        .setName(String.format("Customer %d", i))
+                        .build()
+                )
+                .collect(Collectors.toList());
+        List<TestRecordsJoinIndexProto.OrderWithHeader> orders = IntStream.range(0, customers.size())
+                .mapToObj(i -> TestRecordsJoinIndexProto.OrderWithHeader.newBuilder()
+                        .setHeader(TestRecordsJoinIndexProto.Header.newBuilder().setZKey(1L).setRecId(String.format("order_%d", i)))
+                        .setOrderNo(1000 + i)
+                        .setQuantity(100)
+                        .addAllCc(IntStream.range(0, i).mapToObj(refId -> TestRecordsJoinIndexProto.Ref.newBuilder().setStringValue(String.format("i:%d", refId)).build()).collect(Collectors.toList()))
+                        .build())
+                .collect(Collectors.toList());
+
+        try (FDBRecordContext context = openContext()) {
+            final FDBRecordStore recordStore = recordStoreBuilder.setContext(context).create();
+
+            final SyntheticRecordPlanner planner = new SyntheticRecordPlanner(recordStore);
+            JoinedRecordType joinedRecordType = (JoinedRecordType) recordStore.getRecordMetaData().getSyntheticRecordType(joined.getName());
+            assertConstituentPlansMatch(planner, joinedRecordType, Map.of(
+                    "order",
+                    SyntheticPlanMatchers.joinedRecord(List.of(
+                            PlanMatchers.inComparand(PlanMatchers.hasTypelessString("wrap_int^-1($_j2)"),
+                                    PlanMatchers.typeFilter(Matchers.contains("CustomerWithHeader"),
+                                            PlanMatchers.scan(PlanMatchers.bounds(PlanMatchers.hasTupleString("[EQUALS $_j1, EQUALS $__in_int_rec_id__0]"))))))),
+                    "cust",
+                    // This should be able to use the OrderWithHeader$cc index, but for some reason, is favoring a full scan
```

Comment:
I looked into this, and I believe the problem is #1876 (and not anything to do with joins, *per se*). I have a proposed fix in #1877, though it deals with nested repeated structures, so there may be dragons I'm missing. (That PR is also built on top of this one so that it can change the plan in this test.) AFAICT, this is only a problem with the old planner, though perhaps it needs more testing to verify that.